packages/tools/devtools/devtools-browser-extension/src/content/ContentScript.ts

```diff
@@ -90,4 +95,27 @@ browser.runtime.onConnect.addListener((backgroundPort: Port) => {
 	});
 });
 
+// #region TEST ONLY
+// TODO: compile this out in non-dev builds
+// eslint-disable-next-line @typescript-eslint/no-misused-promises
+window.addEventListener("message", async (event) => {
+	const message = event.data as TestGetTabIdMessage;
+	if (message.type === "TEST_GET_TAB_ID") {
+		console.log("CONTENT SCRIPT: Forwarding tab request from page to background script.");
+
+		// Forward message to background script
+		await browser.runtime.sendMessage(message);
+	}
+});
+
+// Handle TAB_ID message from background script
+browser.runtime.onMessage.addListener((message: TestTabIdMessage) => {
+	if (message.type === "TEST_TAB_ID") {
+		console.log("CONTENT SCRIPT: Forwarding tab ID from background script to webpage.");
+		window?.postMessage(message); // Sent to Application
+	}
+});
```

Comment:
I'm confused here... I would have imagined that the extension would be the one sending the `TestGetTabIdMessage`, so once `ContentScript` receives it, it doesn't need to send it back to the `BackgroundScript` but pass it to devtools/devtools-core. And the handler in `ContentScript` for `TestTabIdMessage` would forward it to `BackgroundScript` (on its way to the extension), not the application.

But also, the listeners we set up above (line 74, line 77) should already take care of forwarding messages as necessary, I don't think we need these ones.

Reply:
These are a special hack. They need to be listening **before** our initialization flow occurs. We should discuss in stand-up parking lot 🙂